### ISSUE TITLE
ChunkIterator and generator support

### ISSUE THREAD

**Issue by robations** [fusonic/linq#20]
Generator's `rewind()` can only be [called once](https://secure.php.net/manual/en/generator.rewind.php), but for some reason I have ChunkIterator trying to rewind again:

```
Cannot rewind a generator that was already run

() at vendor/fusonic/linq/src/Fusonic/Linq/Iterator/ChunkIterator.php:92
 FilterIterator->rewind() at n/a:n/a
 IteratorIterator->rewind() at vendor/fusonic/linq/src/Fusonic/Linq/Iterator/ChunkIterator.php:92
 Fusonic\Linq\Iterator\ChunkIterator->rewind() at vendor/fusonic/linq/src/Fusonic/Linq/Linq.php:409
```

I'll try to create a new test to isolate what is going on with generators specifically.

**Comment by davidroth** [comment]
If we want to use generators, we would have to wrap them in InteratorAggregates and Linq needs to hold the ```IteratorAggregate``` instead of the actual iterator.

For example ```LinqHelper::getIteratorOrThrow()``` needs to be rewritten to directly return the iterator-aggregate instead of returning the iterator.

**Comment by davidroth** [comment]
I have pushed a work in progress branch for this here: https://github.com/fusonic/linq/tree/generator_support

Features: Generator support for Chunk and SelectMany (especially SelectMany gets much cleaner and smaller in code)

**Comment by robations** [comment]
Cool. For the record before #19 was merged generators seemed to work okay for the code I was working on. I see that the `Generator` interface extends `Iterator` (although obviously because of the rewind restriction it doesn't obey Liskov S.P.).

My intuition is that you shouldn't *need* to access or iterate through values more than once.

**Comment by davidroth** [comment]
I have upgraded most of the other iterators. Its much cleaner now and should show better memory usage when iterating over large streams (like in your original issue for chunk).

**Comment by davidroth** [comment]
Merged in f28a3ff99e97e0614875839807ea2efc3667be0f
